### ISSUE TITLE
Multiple Mains (and Brackets too)

### ISSUE THREAD

**Issue by martinglass** [pulquero/RotorHazard#73]
Copied from Alec Lunaford's (Lundy FPV) request on issue #393 on RotorHazard/RotorHazard GitHub repo ...
"I would very much appreciate the ability to run brackets natively. I would also appreciate the ability to run double, triple and greater mains. Having a race conclude on a single race is boring a lot of the time and doesn't favor pilot consistency."

https://github.com/RotorHazard/RotorHazard/issues/393#issuecomment-986253351

**Comment by lundyfpv** [comment]
Yeah I would love if brackets were integrated but my guess is that is a pretty big undertaking.

 I do hope Letter Mains with bumps get added though. Those are pretty essential in my opinion. 
 
Here is a common way I set the mains up in LiveTime, this is a pretty typical way we finish our local races. I change the number of mains, bumps and pilots depending on attendance. 
![image](https://user-images.githubusercontent.com/8151341/144921457-4eec30ab-2662-4a76-a580-568cc77a7346.png)

**Comment by pulquero** [comment]
Please provide details of these formats. I'm not sure I know all the ins-and-outs.

**Comment by lundyfpv** [comment]
Absolutely! So in the picture above that is from LiveTime (LT) and it is how I set the mains to be ran. Now LT allows you to do a lot of customization on how you owuld like to structure your mains. You could run all single mains or you could structure it as above. The way I was doing it above was we had 8 pilots in attendance, the bottom main (D) was a pilot that left early and LT requires me to include registered pilots in mains so disregard the D main. We actually start our mains with C main, this main takes the 4 lowest pilots from qualifying (whatever method you use to seed) and we ran 2 rounds. After the 2 rounds are completed the top 2 pilots from the C main will "bump" to the B main which will then have 3 pilots who will race for 2 rounds. The top 2 pilots after both rounds have completed will bump to the A main with will run 3 rounds before declaring a winner. 

Here is another way of showing the mains as set above:
![image](https://user-images.githubusercontent.com/8151341/146042241-418da67c-c00d-4475-a892-4cf15af87c8d.png)

*Special note here, LiveTime actually weighs the last round in mains higher than the previous rounds. Pilot A gets 1,1,3 will be beaten by Pilot B 2,2,3. I think, but am not 100% that LT weighs the last round 50% more. 

Another way of showing this is a sheet we use for sim races that allows us a lot of flexibility in how we run mains. If you'd like to take a look at the code, I'd be glad to add you as an editor just send me your email. 
https://docs.google.com/spreadsheets/d/1U6-B7nGry4FX7K7UJprQMUBox_Lsb5Y4ITz1V0Gf4iA/edit?usp=sharing

Either way I hope this explains how Mains with Bumps work. 
-Lundy

**Comment by pulquero** [comment]
As a warm-up exercise, I've just added multigp double elimination brackets. I've taken a copy of your sheet, I'll have a look at the formulas to see if I can add this variation too.

Under-the-hood, it is implemented as micro-web-services so it is easy to add more, including third-party ones (nodejs, php, etc).

**Comment by lundyfpv** [comment]
Oh man brackets are awesome! Excited to try this out!

**Comment by pulquero** [comment]
Go into the extended menu, click on race explorer, click on event tab. There is a bracket field for multigp brackets to select which bracket to generate. Workflow would be something like this:
create a qualifying class, add some heats, e.g. generate with randomize, run the heats.
create a new stage and a bracket 1 class, generate bracket 1, run the races.
create a new stage and a bracket 2 class, generate bracket 2, run the races,
etc

*note: for some things, still need to use the old UI.
